components/gatekeeper/auth/AuthServer.go

```diff
@@ -0,0 +1,187 @@
+// Copyright 2019 The Kubeflow Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auth
+
+import (
+	"encoding/base64"
+	"fmt"
+	"github.com/kubeflow/kubeflow/components/gatekeeper/cmd/gatekeeper/options"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/crypto/bcrypt"
+	"math/rand"
+	"net/http"
+	"path"
+	"strings"
+	"sync"
+	"time"
+)
+
+// start with easy case: each server struct only has one valid pair of u/p
+type authServer struct {
+	username	string
+	// password bcrypt hash
+	pwhash      string
+	// authorized cookies and their expire time (12 hour by default)
+	cookies 	map[string]time.Time
+	serverMux   sync.Mutex
+}
+
+const CookieName = "KUBEFLOW-AUTH-KEY"
+const LoginPagePath = "kflogin"
+const LoginPageHeader = "x-from-login"
+
+func NewAuthServer(opt *options.ServerOption) *authServer {
+	data, err := base64.StdEncoding.DecodeString(opt.Pwhash)
+	if err != nil {
+		log.Fatal("error:", err)
+	}
+	server := &authServer{
+		username: opt.Username,
+		pwhash: string(data),
+		cookies: make(map[string]time.Time),
+	}
+	return server
+}
+
+// Default auth check service
+func (s *authServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	log.Infof("Path check, url: %v, path: %v", r.URL, r.URL.Path)
+	// login page open to everyone; all other path requires auth with Password or cookie
+	if strings.HasPrefix(r.URL.Path, "/" + LoginPagePath) || s.authCookie(r) == true {
```

Comment:
Should we be doing an exact match here to path
r.URL.Path, "/" + LoginPagePath

What if by some coincedence we have a path for which this is a prefix but is not an exact match?

Reply:
Browser need access resources like "/LoginPagePath/static/..." as well.
How about do exact match "/LoginPagePath" or prefix match "/LoginPagePath/"?

Reply:
Should be safe to assume any thing behind "/LoginPagePath/" is public?